### PR TITLE
GH-39843: [C++][Parquet] Parquet binary length overflow exception should contain the length of binary

### LINKED ISSUE #39843: [C++][Parquet] Parquet binary length overflow exception should contain the length of binary
### Describe the enhancement requested

```
  template <typename ArrayType>
  void PutBinaryArray(const ArrayType& array) {
    PARQUET_THROW_NOT_OK(::arrow::VisitArraySpanInline<typename ArrayType::TypeClass>(
        *array.data(),
        [&](::std::string_view view) {
          if (ARROW_PREDICT_FALSE(view.size() > kMaxByteArraySize)) {
            return Status::Invalid("Parquet cannot store strings with size 2GB or more");
          }
          PutByteArray(view.data(), static_cast<uint32_t>(view.size()));
          return Status::OK();
        },
        []() { return Status::OK(); }));
  }
```

The code above works well, however, I think it's better to have a length for `view`, which would help debugging ( especially when debugging memory problem, a bad string might having a uninit memory and weird string length)

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

See https://github.com/apache/arrow/issues/39843

It will be great to contain a string length in decoder.

### What changes are included in this PR?

change the logging of encoding

### Are these changes tested?

no

### Are there any user-facing changes?

more specific error logging?

* Closes: #39843

### CODE REVIEW

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -2411,7 +2414,11 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
   void SetData(int num_values, const uint8_t* data, int len) override {
     // num_values is equal to page's num_values, including null values in this page
     this->num_values_ = num_values;
-    decoder_ = std::make_shared<::arrow::bit_util::BitReader>(data, len);
+    if (decoder_ == nullptr) {
```
Comment (wgtmac) [inline cpp/src/parquet/encoding.cc:2417]: What about a separate patch for this?

**Review by wgtmac — APPROVED** [review]

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -160,7 +160,8 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
         *array.data(),
         [&](::std::string_view view) {
           if (ARROW_PREDICT_FALSE(view.size() > kMaxByteArraySize)) {
-            return Status::Invalid("Parquet cannot store strings with size 2GB or more");
+            return Status::Invalid(
+                "Parquet cannot store strings with size 2GB or more, get: ", view.size());
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:164]:
```suggestion
                "Parquet cannot store strings with size 2GB or more, got: ", view.size());
```

Comment (pitrou) [reply cpp/src/parquet/encoding.cc:164]: (same comment for other occurrences below)

**Comment by mapleFU** [comment]
comment fixed @pitrou

**Review by pitrou — APPROVED** [review]
Feel free to merge if CI is ok :-)

**Comment by mapleFU** [comment]
Don't know why CI continue fails. Rebase and run ci again

**Comment by mapleFU** [comment]
I've checked the CI, seems these test failed for a few days. I'll merge it firstly

**Comment by mapleFU** [comment]
Why do conbench has error =_=

**Pull request merged**